### PR TITLE
rgw: awsv4 client support

### PR DESCRIPTION
**Pull request by yehudasa**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

Provides support for the rest client to use the more secure aws v4 signatures. There is a daemon configurable that can be used to revert to v2 signatures.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/iso_8601.cc [hunk 1]
```diff
@@ -199,4 +199,54 @@ string to_iso_8601(const real_time t,
   out << '.' << setw(9) << nsec << 'Z';
   return out.str();
 }
+
+string to_iso_8601_no_separators(const real_time t,
+                                 const iso_8601_format f) noexcept {
+  ceph_assert(f >= iso_8601_format::Y &&
```
Comment (mattbenjamin) [inline src/common/iso_8601.cc:205]: should we be calling ceph_assert() here?  is this intended to be a compile-time check?

File: src/common/iso_8601.cc [hunk 2]
```diff
@@ -199,4 +199,54 @@ string to_iso_8601(const real_time t,
...
+	      f <= iso_8601_format::YMDhmsn);
+  stringstream out(std::ios_base::out);
```
Comment (mattbenjamin) [inline src/common/iso_8601.cc:207]: it's easy to replace later, but I think everything done here could be handled by fmt more efficiently

**Review by mattbenjamin — COMMENTED** [review]
initial comments

File: src/rgw/rgw_http_client.cc [hunk 3]
```diff
@@ -294,6 +294,51 @@ void RGWIOProvider::assign_io(RGWIOIDProvider& io_id_provider, int io_type)
...
+      verify_ssl(cct->_conf->rgw_verify_ssl),
+      cct(cct),
+      method(_method),
+      url(_url) {
+  init();
+}
+
+std::ostream& RGWHTTPClient::gen_prefix(std::ostream& out) const
+{
+  out << "http_client[" << method << "/" << url << "]";
```
Comment (cbodley) [inline src/rgw/rgw_http_client.cc:313]: this is a neat idea! maybe `method << " " << url` so we see `GET /admin/log`?

using keystone auth as an example, we're running RGWHTTPClient in the context of a request, so we want to see the prefix from its req_state and op too

i'd suggest using `DoutPrefixPipe` instead, so you can wrap the caller's dpp and use `add_prefix()` to add this extra output

you should also update the `dout()`s to `ldpp_dout()`s in the class so we actually see this

**Review by cbodley — COMMENTED** [review]
👍 👍

File: src/rgw/rgw_zone.h [hunk 4]
```diff
@@ -841,6 +841,7 @@ struct RGWPeriodMap
   std::string id;
   map<std::string, RGWZoneGroup> zonegroups;
   map<std::string, RGWZoneGroup> zonegroups_by_api;
+  map<rgw_zone_id, std::shared_ptr<RGWZoneGroup> > zonegroups_by_zone;
```
[the remark this answers is not in the document]
Comment (yehudasa) [reply src/rgw/rgw_zone.h:844]: yeah, unused ... probably leftovers from the original branch. I'll remove

File: src/rgw/rgw_rest_s3.h [hunk 5]
```diff
@@ -1079,6 +1079,48 @@ class AWSBrowserUploadAbstractor : public AWSEngine::VersionAbstractor {
   auth_data_t get_auth_data(const req_state* s) const override;
 };
 
+class AWSSignerV4 {
+  const DoutPrefixProvider *dpp;
+  CephContext *cct;
+
+public:
+  AWSSignerV4(const DoutPrefixProvider *_dpp) : dpp(_dpp),
```
[the remark this answers is not in the document]
Comment (yehudasa) [reply src/rgw/rgw_rest_s3.h:1087]: ok

File: src/rgw/rgw_rest_s3.cc [hunk 6]
```diff
@@ -5204,6 +5204,144 @@ AWSGeneralAbstractor::get_v4_canonical_headers(
...
+    extra_headers["x-amz-content-sha256"] = content_hash;
+
+  }
+
+  /* craft canonical headers */
+  std::string canonical_headers = \
+    gen_v4_canonical_headers(info, extra_headers, &signed_hdrs);
+
+  using sanitize = rgw::crypt_sanitize::log_content;
+  ldout(cct, 10) << "canonical headers format = "
```
[the remark this answers is not in the document]
Comment (yehudasa) [reply src/rgw/rgw_rest_s3.cc:5243]: sure

File: src/common/iso_8601.cc (hunk 2 above)
Comment (yehudasa) [reply src/common/iso_8601.cc:207]: yeah, I ended implementing casey's suggestion (to reuse `to_iso_8601()`)

File: src/rgw/rgw_auth_s3.cc [hunk 7]
```diff
@@ -491,6 +491,30 @@ int parse_v4_credentials(const req_info& info,                     /* in */
...
+
+  return out.str();
+}
+
+string gen_v4_scope(const ceph::real_time& timestamp,
+                    const string& region,
+                    const string& service)
+{
+
+  return to_scope_date(timestamp) + "/" + region + "/" + service + "/aws4_request";
```
[the remark this answers is not in the document]
Comment (yehudasa) [reply src/rgw/rgw_auth_s3.cc]: done

File: src/common/iso_8601.cc (hunk 2 above)
Comment (mattbenjamin) [reply src/common/iso_8601.cc]: cool

**Comment by yehudasa** [comment]
https://pulpito.ceph.com/yehuda-2021-03-11_22:59:05-rgw-wip-rgw-aws4-client-distro-basic-smithi/

**Pull request merged**